src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/HashPartitioningItemMapper.java

```diff
@@ -0,0 +1,359 @@
+            switch (type) {
+                case S:
+                    return value.getS().getBytes(StandardCharsets.UTF_8);
+                case N:
+                    BigDecimal bigDecimal = new BigDecimal(value.getN());
+                    return BigDecimalSortedBytesConverter.encode(bigDecimal);
+                case B:
+                    return value.getB().array();
+                default:
+                    throw new UnsupportedOperationException("Unsupported field type: " + type);
+            }
+        }
+
+        static AttributeValue[] fromBytes(ScalarAttributeType hkType, ScalarAttributeType rkType, ByteBuffer buf) {
+            short hkValueLength = buf.getShort();
+            AttributeValue hkValue = fromBytes(hkType, buf, hkValueLength);
+            AttributeValue rkValue = fromBytes(rkType, buf, buf.remaining());
+            return new AttributeValue[] { hkValue, rkValue };
+        }
+
+        static AttributeValue fromBytes(ScalarAttributeType type, ByteBuffer buf) {
+            return fromBytes(type, buf, buf.limit());
+        }
+
+        private static AttributeValue fromBytes(ScalarAttributeType type, ByteBuffer buf, int size) {
+            byte[] bytes = new byte[size];
+            buf.get(bytes);
+            switch (type) {
+                case S:
+                    return new AttributeValue(new String(bytes, StandardCharsets.UTF_8));
+                case N:
+                    return new AttributeValue().withN(BigDecimalSortedBytesConverter.decode(bytes).toPlainString());
+                case B:
+                    return new AttributeValue().withB(ByteBuffer.wrap(bytes));
+                default:
+                    throw new UnsupportedOperationException("Unsupported field type: " + type);
+            }
+        }
+    }
+
+    /**
+     * Encodes BigDecimals into byte arrays such that the numerical ordering is preserved.
+     * <p/>
+     * In Dynamo, a number can have up to 38 digits of precision, and can be positive, negative, or zero.
+     * <p><ul>
+     * <li>Positive range: 1E-130 to 9.9999999999999999999999999999999999999E+125
+     * <li>Negative range: -9.9999999999999999999999999999999999999E+125 to -1E-130
+     * </ul></p>
+     * (see doc: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-data-types)
+     * <p/>
+     * The format we use is SIGNUM + EXPONENT + D_1 + ... + D_n, where
+     * <p><ul>
+     * <li>SIGNUM: -1, 0, or 1, corresponding to negative, zero, or positive (1 byte)
+     * <li>EXPONENT: Represents exponent when number is in normalized notation (range is [-128, 127], mapped to the
+     * Dynamo range [-130, 125]; 1 byte)
+     * <li>D_1 to D_n: The n (at most 38) significand digits, from most significant to least (each digit takes 1 byte)
+     * <li>When the sign is negative, the exponent and significand digits' bits are flipped
+     * </ul></p>
+     */
+    @VisibleForTesting
```

Comment:
Once numbers are put in normalized scientific notation, it's straightforward to sort them. E.g., it's discussed here: https://softwareengineering.stackexchange.com/questions/246021/trivially-sortable-encoding-for-arbitrary-precision-decimals.

I also looked at [how HBase does it](https://hbase.apache.org/apidocs/org/apache/hadoop/hbase/util/OrderedBytes.html), which in turn is basically [how SQLite does it](https://sqlite.org/src4/doc/trunk/www/key_encoding.wiki), but their math is a bit more complicated, and I don't know why the math works (though I'm sure Jeremy can figure it out easily).